### PR TITLE
Add LOG015 to detect use of the root logger

### PR DESCRIPTION
**Pull request by jakkdl**
The implementation of the check itself is super simple. But unsurprisingly it gives errors on tons of the examples on the other test checks. I first resolved it in `TestLOG003` by actually adding the errors, but as you can see in the first commit that's not really tenable - so I instead added a way to ignore errors when running tests.

None of the `TestLOG009` cases gives `LOG015` since it only checks the calls in `logger_methods`. I don't think it's worth being that thorough though, but would be straightforward to extend the check.

It also remains a decision point on whether this check should be enabled by default or not, if you want me to make it ignored by default I can implement that as well.

### CODE REVIEW

**Comment by jakkdl** [comment]
Fixes #87

**Comment by adamchainz** [comment]
Please add a changelog note.

**Comment by Zac-HD** [comment]
As a complementary check, I'd suggest expanding `LOG002` to warn on `None`.

**Comment by jakkdl** [comment]
> As a complementary check, I'd suggest expanding `LOG002` to warn on `None`.

I think that depends on whether `logging.root` or `logging.getLogger(None)` should be considered the canonical way of explicitly saying you want the root logger (and that explicitly wanting the root logger is a valid thing to want in a non-trivial amount of cases). As noted by adamchainz in #87 `logging.root` does not seem to be mentioned in [the online docs](https://docs.python.org/3/library/logging.html#module-logging) at all, only place I can find a mention of it is when calling `help(logging.Logger)`:
```
...

 |  ----------------------------------------------------------------------
 |  Data and other attributes defined here:
 |  
 |  manager = <logging.Manager object>
 |  
 |  root = <RootLogger root (WARNING)>
 |  
 |  ----------------------------------------------------------------------

...
```

So idk if we should recommend it? Or maybe this is just a case of the online docs being subpar and there being 0 risk that it gets deprecated or anything in the future.

If wanting the root logger is bad 99% of the time then outlawing ~all ways of accessing it and requiring `# noqa` to silence the warning seems fine, but otherwise we probably want a way to explicitly request it that doesn't raise any warnings.

Of course also an option to add a separate opt-in check for `logging.getLogger(None)` and/or `logging.root`

**Comment by jakkdl** [comment]
> Please add a changelog note.

done! :heavy_check_mark:

File: CHANGELOG.rst [hunk 1]
```diff
@@ -2,6 +2,14 @@
 Changelog
 =========
 
+1.6.0 (2024-03-20)
+------------------
+
```
Comment (adamchainz) [inline CHANGELOG.rst:7]: added at release time

```suggestion
```

File: CHANGELOG.rst [hunk 2]
```diff
@@ -2,6 +2,14 @@
...
+* Add rule LOG015 that detects logger calls directly on the root logger ``logging``.
```
Comment (adamchainz) [inline CHANGELOG.rst:8]:
```suggestion
* Add rule LOG015 that detects use of the root logger through calls like ``logging.info()``.
```

File: README.rst [hunk 3]
```diff
@@ -563,3 +563,27 @@ Corrected:
     import logging
 
     logging.warning("Uh oh")
+
```
Comment (adamchainz) [inline README.rst:566]: one line only
```suggestion
```

File: README.rst [hunk 4]
```diff
@@ -563,3 +563,27 @@ Corrected:
...
+
+LOG015 avoid logging calls on root logger
+-----------------------------------------
+
+Making logging calls directly on the root logger makes it hard to later find the origin when reading logs. You should instead instantiate a logger object with e.g. ``logging.getLogger(__name__)``. If you really do want to interface with the root object you should explicitly use ``logging.root``.
```
Comment (adamchainz) [inline README.rst:571]:
```suggestion
Using the root logger means your messages have no source information, making them less useful for debugging.
It’s better to always create a logger object, normally with:

.. code-block:: python
    
    logger = logging.getLogger(__name__)

If you really do need the root logger, use ``logging.getLogger(None)``.

This rule detects any call to a logging method directly on the ``logging`` module object.
```

File: src/flake8_logging/__init__.py [hunk 5]
```diff
@@ -128,6 +128,7 @@ def modnamed_placeholder_re() -> re.Pattern[str]:
 LOG012 = "LOG012 formatting error: {n} {style} placeholder{ns} but {m} argument{ms}"
 LOG013 = "LOG013 formatting error: {mistake} key{ns}: {keys}"
 LOG014 = "LOG014 avoid exc_info=True outside of exception handlers"
+LOG015 = "LOG015 avoid logging calls on root logger"
```
Comment (adamchainz) [inline src/flake8_logging/__init__.py:131]:
```suggestion
LOG015 = "LOG015 avoid logging calls on the root logger"
```

File: README.rst [hunk 6]
```diff
@@ -563,3 +563,27 @@ Corrected:
     import logging
 
     logging.warning("Uh oh")
+
+
+LOG015 avoid logging calls on root logger
+-----------------------------------------
```
Comment (adamchainz) [inline README.rst:569]:
```suggestion
LOG015 avoid logging calls on the root logger
---------------------------------------------
```

**Review by adamchainz — COMMENTED** [review]
Gonna fix up and merge this now, some review with changes that I will apply.

**Comment by adamchainz** [comment]
Released in 1.6.0. Great work, thank you!

**Comment by jakkdl** [comment]
Thank you for thorough review & cleanup!

**Pull request merged**